pkg/clusteragent/clusterchecks/dispatcher_rebalance.go

```diff
@@ -0,0 +1,208 @@
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-2019 Datadog, Inc.
+
+// +build clusterchecks
+
+package clusterchecks
+
+import (
+	"fmt"
+	"sort"
+
+	"github.com/DataDog/datadog-agent/pkg/util/log"
+)
+
+type Weight struct {
+	nodeName string
+	busyness int
+}
+
+type Weights []Weight
+
+func (w Weights) Len() int           { return len(w) }
+func (w Weights) Less(i, j int) bool { return w[i].busyness > w[j].busyness }
+func (w Weights) Swap(i, j int)      { w[i], w[j] = w[j], w[i] }
+
+func (d *dispatcher) calculateAvg() (int, error) {
+	busyness := 0.0
+	length := 0
+
+	d.store.RLock()
+	defer d.store.RUnlock()
+
+	for _, node := range d.store.nodes {
+		node.RLock()
+		for _, stats := range node.clcRunnerStats {
+			busyness += busynessFunc(stats.AverageExecutionTime, stats.MetricSamples)
+		}
+		length += 1
+		node.RUnlock()
+	}
+
+	if length == 0 {
+		return -1, fmt.Errorf("zero nodes reporting")
+	}
+
+	return int(busyness) / length, nil
+}
+
+// getDiffAndWeights creates a map that contains the difference between
+// the busyness on each node and the total average busyness, and a Weights
+// struct containing nodes and their busyness values
+func (d *dispatcher) getDiffAndWeights(avg int) (map[string]int, Weights) {
+	diffMap := make(map[string]int)
+	weights := Weights{}
+
+	d.store.RLock()
+	defer d.store.RUnlock()
+
+	for nodeName, node := range d.store.nodes {
+		node.RLock()
+		busyness := calculateBusyness(node.clcRunnerStats)
+		node.RUnlock()
+		diffMap[nodeName] = busyness - avg
+		weights = append(weights, Weight{
+			nodeName: nodeName,
+			busyness: busyness,
+		})
+	}
+	return diffMap, weights
+}
+
+// updateDiff creates a map that contains the difference between
+// the busyness on each node and the total average busyness.
+func (d *dispatcher) updateDiff(avg int) map[string]int {
+	diffMap := make(map[string]int)
+
+	d.store.RLock()
+	defer d.store.RUnlock()
+
+	for nodeName, node := range d.store.nodes {
+		node.RLock()
+		busyness := calculateBusyness(node.clcRunnerStats)
+		node.RUnlock()
+		diffMap[nodeName] = busyness - avg
+	}
+	return diffMap
+}
+
+// pickCheckToMove select the most appropriate check to move from a node to another.
+// A check Xi running on a node N is chosen to move to another node if it satisfies the following
+// Weight(Xi) >  Weight(Xj) (for each j != i, 0 <= j < len(weights))
+// where Weight(X) is the busyness value caused by running the check X.
+func (d *dispatcher) pickCheckToMove(nodeName string) (string, int, error) {
+	(*d.store.nodes[nodeName]).RLock()
+	defer (*d.store.nodes[nodeName]).RUnlock()
+
+	if len((*d.store.nodes[nodeName]).clcRunnerStats) == 0 {
+		return "", -1, fmt.Errorf("Node %s has no check stats", nodeName)
+	}
+
+	firstItr := true
+	checkID := ""
+	checkWeight := 0
+	for id, stats := range (*d.store.nodes[nodeName]).clcRunnerStats {
+		busyness := int(busynessFunc(stats.AverageExecutionTime, stats.MetricSamples))
```

Comment:
same logic than ```calculateBusyness()``` ?

Reply:
if we agree on moving the lock to the node store, yes we can follow the same logic here